rpc/common.go

```diff
@@ -133,30 +133,36 @@ func GetGroupControllerCapabilities(ctx context.Context, conn *grpc.ClientConn)
 
 // ProbeForever calls Probe() of a CSI driver and waits until the driver becomes ready.
 // Any error other than timeout is returned.
-func ProbeForever(conn *grpc.ClientConn, singleProbeTimeout time.Duration) error {
+func ProbeForever(ctx context.Context, conn *grpc.ClientConn, singleProbeTimeout time.Duration) error {
+	logger := klog.FromContext(ctx)
+	ticker := time.NewTicker(probeInterval)
+
 	for {
-		klog.Info("Probing CSI driver for readiness")
-		ready, err := probeOnce(conn, singleProbeTimeout)
-		if err != nil {
-			st, ok := status.FromError(err)
-			if !ok {
-				// This is not gRPC error. The probe must have failed before gRPC
-				// method was called, otherwise we would get gRPC error.
-				return fmt.Errorf("CSI driver probe failed: %s", err)
-			}
-			if st.Code() != codes.DeadlineExceeded {
-				return fmt.Errorf("CSI driver probe failed: %s", err)
-			}
-			// Timeout -> driver is not ready. Fall through to sleep() below.
-			klog.Warning("CSI driver probe timed out")
-		} else {
-			if ready {
-				return nil
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
The original code did not return on timeout, it kept trying. I think you should return an error here to indicate that probing timed out.

```suggestion
			return ctx.Err()
```

Reply:
https://github.com/kubernetes-csi/csi-lib-utils/pull/149/commits/d0c5c66abd0ac95a8f5d62e2b36a3beaa3cc5f59#diff-5fb7c6e20b635d2fb705b5f7ec97d8f1a35c802d9bd76105b814b43ee4885c6eL143-R143